include/AIToolbox/MDP/Algorithms/MCTS.hpp

```diff
@@ -43,20 +42,22 @@ namespace AIToolbox::MDP {
      * Then it simply makes that root branch the new root, and starts
      * again.
      */
-    template <typename M>
+    template <typename M, typename Sel, typename StateT = size_t, typename ActionT = size_t>
     class MCTS {
-        static_assert(is_generative_model_v<M>, "This class only works for generative MDP models!");
+        static_assert(is_generative_model_v<M, StateT, ActionT>, "This class only works for generative MDP models!");
 
         public:
+            using RandomEngine = std::mt19937;
```

Comment:
I understand the reasoning behind wanting to decouple MCTS from Eigen, but by doing this you are basically decoupling MCTS from the entire library. And at the same time, it's not reasonable for me to maintain separate headers introducing typedefs for any particular dependency I have. It makes life easier for people that want to take out things from the library, but makes stuff much harder for me to track dependencies and duplications. So I can't really get behind this.

Reply:
That's fair enough but at the same time it puts a higher entry cost on people wanting to use only parts of your library. Is a compilation flag like '-DNO_EIGEN' that defaults to the current behaviour (i.e., having it required) so that it wouldn't change anything for people who don't care better in your opinion?

Reply:
I think this would basically result in the same complexity, no? What would the NO_EIGEN symbol do? I'd need to have all typedefs wrapped into `#ifdefs`, and again I would need to test all code against the symbol being defined or not. The CMake would need to selectively include files which can be included as not dependent on Eigen, and by extension I'd have to keep track of which files I can or cannot include every time I write something. I feel that would be really lots of work, and complicate things for me quite a lot. And again, where do you stop? You're now doing it for MCTS. How many other methods could really be decoupled from Eigen? In the code there's even multiple implementations for some methods (for example ValueIteration) that can work whether you define a model with Eigen or not. Should we decouple them as well, so that you can compile them separately as well? I feel it would become a mess very very quickly.

In general, I don't think asking an user to build the project is too much to ask. That's how software works. The tests passing that it's possible to do, and so I wouldn't really want to have lines of code supporting stuff I don't need & having the responsibility of tracking dependencies until the end of times.

Reply:
That's because the travis is on a Debian. For instance, I build most of my stuff on Alpine, and lpsolve does not seem to exist. So we can argue that anybody using Alpine can build libs from source but it also means that more dependencies means more tracking of them **for the user** because they will need to update them themselves. However, I can relate with all your rational. I'll think about that.

Reply:
Fair. On the other hand, the user also saves tons of time since they are using the library and not having to re-implement MCTS or what have you from scratch ;) So I still think it's fair to require users to build their own dependencies.

Again, I see what you are coming from. I did try to reduce the required dependencies in the past. For example, at the beginning `lp_solve` was only required for POMDPs, which is pretty much your point. Eigen was introduced after some time, as I used to use vectors only for storage. This has changed over time both because of new features but also because it's easier for me, and I'm just one guy here (you're the first serious contributor in like 5 years). In the end, supporting too many things in this way is just not worth it to me. The alternative is to mess up the code, mess up the build system, test a million possible combinations of flags (which is something I have specifically designed against throughout, from builds to algo options).

Reply:
After a night sleep, I am backing you 100. I have an idea that would mean only have one ```#ifndef``` in ```Types``` and shadowing the ```Matrix2D``` and stuff types but I'll leave that for the future. For now I'll revert the changes. In my case, because I'll use only the headers, it only means adding ```Eigen-dev``` that is in Alpine so it's perfectly fine.